### DISCUSSION TITLE
API Keys and Secrets in headers using httpc

### DISCUSSION THREAD

**Discussion by rosssaunders** [omnigres/omnigres#377] [Q&A]
Hi,

I've been playing around with the httpc extension to call out to REST endpoints. The majority of the endpoints I'm calling require an API key, secret or both in the headers. Given that Postgres can log all SQL statements, this would mean that the secrets are then available to anyone with admin access.

I've checked the omnigres documentation and couldn't see any references on how to ensure that secrets are not logged by Postgres. Are there any plans to make the HTTPC extension flexible so that it could load secrets at that point so it would not available to Postgres?

Thanks!

**Comment by yrashk** [comment]
Great question. Ultimately, if you're concerned about admins, I don't think there's a fool-proof way to protect. Admins can replace any function with whatever they'd like.

That said, enclosing a call into a function can be useful:

```
# create function a() returns int language sql as $$ select count(*) from pg_proc $$;
2023-11-21 13:52:10.286 PST [85732] LOG:  statement: create function a() returns int language sql as $$ select count(*) from pg_proc $$;
CREATE FUNCTION
Time: 2.091 ms
omni_httpc=# select a();
2023-11-21 13:52:15.900 PST [85732] LOG:  statement: select a();
  a   
------
 3298
(1 row)
```

The above didn't divulge the query inside of a function.

Some Omnigres users have a credentials table that is only available to certain roles and wrap their calls to a particular endpoint/service into a function that is declared as `security definer` and is owned by a permitted role. This way normal role won't have access to these credentials.

However, the administrators are always able to see these credentials.

Am I getting your case right?

**Reply by rosssaunders** [reply]
Thanks for the quick response. I'll do some more playing around with wrapping the calls as functions to confirm what is potentially be able to be logged. I'll let you know.
